weave-js/src/components/PagePanelComponents/Home/Browse3/pages/wfReactInterface/tsDataModelHooksCallRefExpansion.ts

```diff
@@ -71,98 +82,350 @@ export const useClientSideCallRefExpansion = (
-          value = null;
-          break;
+        if (isExpandableRef(value)) {
+          refsNeeded.add(value);
         }
-        value = value[part];
-      }
-      while (
-        typeof value === 'object' &&
-        value != null &&
-        EXPANDED_REF_VAL_KEY in value
-      ) {
-        value = value[EXPANDED_REF_VAL_KEY];
-      }
-      if (isExpandableRef(value)) {
-        refsNeeded.add(value);
-      }
+      });
     });
-  });
+  }
 
   const refsNeededArray = Array.from(refsNeeded);
 
   if (refsNeededArray.length === 0) {
     return traceCalls;
   }
 
-  const refsData = await directFetchRefsData(refsNeededArray, client);
-  const refsDataMap = new Map<string, any>();
-  refsNeededArray.forEach((ref, i) => {
-    refsDataMap.set(ref, refsData[i]);
-  });
+  try {
+    const refsData = await directFetchRefsData(refsNeededArray, client);
+    const refsDataMap = new Map<string, any>();
+    refsNeededArray.forEach((ref, i) => {
+      refsDataMap.set(ref, refsData[i]);
+    });
 
-  const expandedTraceCalls = traceCalls.map(call => {
-    call = _.cloneDeep(call);
-    expandedRefColumnsList.forEach(col => {
-      const colParts = col.split('.');
-      let value: any = call;
-      const path: string[] = [];
-      for (const part of colParts) {
-        while (
-          typeof value === 'object' &&
-          value != null &&
-          EXPANDED_REF_VAL_KEY in value
-        ) {
-          value = value[EXPANDED_REF_VAL_KEY];
-          path.push(EXPANDED_REF_VAL_KEY);
-        }
-        if (value == null) {
-          break;
-        }
-        if (typeof value !== 'object' || !(part in value)) {
-          value = null;
-          break;
-        }
-        value = value[part];
-        path.push(part);
+    const expandedTraceCalls = traceCalls.map(call => {
+      call = _.cloneDeep(call);
+
+      if (recursiveUnwrap) {
+        // Recursively replace all references, but only in specified paths
+        const expandedRefColumnsList = Array.from(expandedRefColumns ?? []);
+
+        expandedRefColumnsList.forEach(column => {
+          // Navigate to the column path
+          const parts = column.split('.');
+          let target: any = call;
+          let parent: any = null;
+
+          // Find the parent object that contains the target field
+          for (let i = 0; i < parts.length; i++) {
+            const part = parts[i];
+            if (target == null || typeof target !== 'object') {
+              target = undefined;
+              break;
+            }
+
+            if (i === parts.length - 1) {
+              parent = target;
+            }
+
+            target = target[part as keyof typeof target];
+          }
+
+          // If we found the path, recursively replace refs in it
+          if (target !== undefined && parent !== null) {
+            // Replace refs in the target value
+            replaceAllRefsRecursively(target, refsDataMap);
```

Comment:
this resolved all refs for top level provided keys, so `inputs.model` could resolve lots and lots of sub refs right?